pkg/ring/kv/multi.go

```diff
@@ -316,9 +316,13 @@ func (m *MultiClient) keyUpdated(primary kvclient, key string, newValue interfac
 	}
 }
 
-// Stop the multiClient (mirroring part).
-func (m *MultiClient) stop() {
+// Stop the multiClient (mirroring part), and all configured clients.
+func (m *MultiClient) Stop() {
 	m.cancel()
+
+	for _, kv := range m.clients {
+		kv.client.Stop()
```

Comment:
Consul and etcd `Stop()` are no-op, so who is using it right now? Has this been introduced because of the Gossip client? If so, do you need to wait before the `cancel()` takes effect (the on-going operations have been cancelled) or not necessary?

Reply:
This is a new method introduced by Gossiping work. It's no-op in Consul and Etcd (although they could do some cleanup, if necessary), while memberlist client uses it to leave the cluster gracefully... it is especially useful when ingester or distributor wants to remove itself from the ring, before leaving -- this gives gossiping client a chance to send it out.

Cancel only stops the mirroring part [update: and now also watching of the config channel]. It doesn't cancel ongoing operations.